src/psyclone/transformations.py

```diff
@@ -2854,3 +2854,282 @@ def validate(self, loop):
+        :raises TransformationError: if transformation is applied to a Loop \
+                                     without its parent Directive when \
+                                     optimisations are applied.
+        :raises TransformationError: if transformation is applied to an \
+                                     orphaned Directive without its parent \
+                                     Directive.
+        '''
+
+        # First check constraints on Nodes in the node_list common to
+        # all RegionTrans transformations.
+        super(ExtractRegionTrans, self)._validate(node_list)
+
+        # Now check ExtractRegionTrans specific constraints.
+
+        # Extracting distributed memory code is not supported due to
+        # generation of infrastructure calls to set halos dirty or clean.
+        # This constraint covers the presence of HaloExchange and
+        # GlobalSum classses as they are only generated when distributed
+        # memory is enabled.
+        if Config.get().distributed_memory:
+            raise TransformationError(
+                "Error in {0}: Distributed memory is not supported."
+                .format(str(self.name)))
+
+        # Check constraints not covered by valid_node_types for
+        # individual Nodes in node_list.
+        from psyclone.psyGen import Loop, Kern, BuiltIn, Directive, \
+            OMPDoDirective, ACCLoopDirective, OMPParallelDoDirective
+
+        for node in node_list:
+
+            # Check that ExtractNode is not inserted between a Kernel or
+            # a BuiltIn call and its parent Loop.
+            if isinstance(node, (Kern, BuiltIn)) and \
+               isinstance(node.parent, Loop):
+                raise TransformationError(
+                    "Error in {0}: Extraction of a Kernel or a Built-in "
+                    "call without its parent Loop is not allowed."
+                    .format(str(self.name)))
+
+            # Check that ExtractNode is not inserted between a Loop and its
+            # parent Directive when optimisations are applied, as this may
+            # result in including the end of Directive for extraction but
+            # not the beginning.
+            if isinstance(node, Loop) and isinstance(node.parent, Directive):
+                raise TransformationError(
+                    "Error in {0}: Extraction of a Loop without its parent "
+                    "Directive is not allowed.".format(str(self.name)))
+
+            # Check that ExtractNode is not inserted between an orphaned
+            # Directive (e.g. OMPDoDirective, ACCLoopDirective) and its
+            # ancestor Directive (e.g. ACC or OMP Parallel Directive) when
+            # optimisations are applied. Note that we need to explicitly
+            # exclude the OMPParallelDoDirective as it inherits from both
+            # OMPDoDirective and OMPParallelDirective.
+            if isinstance(node, (OMPDoDirective, ACCLoopDirective)) and \
```

Comment:
Upon reflection, I think that this check needs to be a bit more thorough. Currently this code assumes that any orphaned directives are immediate children of the Extract node. In practice, they may be further down than that (I'm thinking here of the NEMO API but it could be/become an issue for Dynamo and GOcean too). I think what we need to do is `node.walk()` down each sub-tree and check for directives. If the first directive we encounter in any of these walks is an orphaned directive then we raise an error.

Reply:
In fact, having just looked at your GOcean-specific sub-class of ExtractTrans I see there's a better way to do it - check on the types of any directives that are ancestors of the supplied list of nodes.

Reply:
Done. Also deleted the test for the orphaned OMP directive and just left the one for the orphaned ACC directive from `extract_test.py. They are testing the same constraint as you commented.

Reply:
Nice - thanks.